testutil/fuse/fuse.go

```diff
@@ -0,0 +1,130 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fuse
+
+import (
+	"fmt"
+	"github.com/prometheus/tsdb/testutil"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/hanwen/go-fuse/fuse"
+	"github.com/hanwen/go-fuse/fuse/nodefs"
+	"github.com/hanwen/go-fuse/fuse/pathfs"
+)
+
+type loopBackFs struct {
+	pathfs.FileSystem
+}
+
+// A FUSE filesystem that shunts all request to an underlying file system.
+// Its main purpose is to provide test coverage.
+type hookFs struct {
+	original   string
+	mountpoint string
+	fsName     string
+	loopBackFs
+	hook Hook
+}
+
+func newHookFs(original string, mountpoint string, hook Hook) (*hookFs, error) {
+	hookfs := &hookFs{
+		original:   original,
+		mountpoint: mountpoint,
+		fsName:     "hookfs",
+		loopBackFs: loopBackFs{pathfs.NewLoopbackFileSystem(original)},
+		hook:       hook,
+	}
+	return hookfs, nil
+}
+
+// String implements hanwen/go-fuse/fuse/pathfs.FileSystem. You are not expected to call h manually.
+func (h *hookFs) String() string {
+	return fmt.Sprintf("HookFs{Original=%s, Mountpoint=%s, FsName=%s, Underlying fs=%s, hook=%s}",
+		h.original, h.mountpoint, h.fsName, h.loopBackFs.String(), h.hook)
+}
+
+// Rename implements hanwen/go-fuse/fuse/pathfs.FileSystem. You are not expected to call h manually.
+func (h *hookFs) Rename(oldName string, newName string, context *fuse.Context) fuse.Status {
+	preHooked, err := h.hook.PreRename(oldName, newName)
+	if preHooked {
+		if err != nil {
+
+			return fuse.ToStatus(err)
+		}
+	}
+
+	status := h.loopBackFs.Rename(oldName, newName, context)
+
+	postHooked, err := h.hook.PostRename(oldName, newName)
+	if postHooked {
+		if err != nil {
+			return fuse.ToStatus(err)
+		}
+	}
+	return status
+}
+
+func (h *hookFs) newServer() (*fuse.Server, error) {
+	opts := &nodefs.Options{
+		NegativeTimeout: time.Second,
+		AttrTimeout:     time.Second,
+		EntryTimeout:    time.Second,
+	}
+	pathFsOpts := &pathfs.PathNodeFsOptions{ClientInodes: true}
+	pathFs := pathfs.NewPathNodeFs(h, pathFsOpts)
+	conn := nodefs.NewFileSystemConnector(pathFs.Root(), opts)
+	originalAbs, _ := filepath.Abs(h.original)
+	mOpts := &fuse.MountOptions{
+		AllowOther: true,
+		Name:       h.fsName,
+		FsName:     originalAbs,
+	}
+	server, err := fuse.NewServer(conn.RawFS(), h.mountpoint, mOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	return server, nil
+}
+
+// NewServer creates a fuse server and attaches it to the given `mountpoint` directory.
+// It returns a function to `unmount` the given `mountpoint` directory.
+func NewServer(t *testing.T, original, mountpoint string, hook Hook) (clean func()) {
+	fs, err := newHookFs(original, mountpoint, hook)
+	testutil.Ok(t, err)
+
+	server, err := fs.newServer()
+	testutil.Ok(t, err)
```

Comment:
I think it is best that when there is an error to just skip the test.
This will not enforce users to have a fuse server installed locally to run tests and we will make sure the these are not skipped during CI.

Or if there is any other way to detect when fuse server is not installed, show a warning and continue with the other tests.

Reply:
There is a fast fail if fuse  is not installed.
I think whether skip the error should be controlled by test case.
So if test case call `fuse.NewServer` and this function return error,  the test case decide to skip it.
Do you think it is ok?

Reply:
I didn't quite understand the, idea, can you maybe show it with some code.

Reply:
With the new code, if the custom do not install fuse locally, the test case will skip tests.
I run it in a machine which do not install fuse locally, the result is here:
```
[root@172.16.4.65 tsdb]# go test -timeout 60s github.com/prometheus/tsdb -run ^TestFailedDelete$ -v

=== RUN   TestFailedDelete
--- SKIP: TestFailedDelete (0.36s)
    compact_test.go:1078: use server couldn't be started
PASS
ok  	github.com/prometheus/tsdb	0.383s
```